### PR TITLE
Add TF whisper

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Adds TF Whisper port of PyTorch implementation

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 1]
```diff
@@ -0,0 +1,866 @@
...
+
+        return config, input_ids, None, max_length
+
+    # not implemented currently
+    def test_inputs_embeds(self):
+        pass
+
+    unittest.skip("Training is not yet supported")
+
+    def test_training(self):
```
Comment (gante) [inline tests/models/whisper/test_modeling_tf_whisper.py:322]:
```suggestion
    @unittest.skip("Training is not yet supported")
    def test_training(self):
```

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 2]
```diff
@@ -0,0 +1,866 @@
...
+    def test_generate_with_head_masking(self):
+        pass
+
+    def test_generate_fp16(self):
+        config, input_dict = self.model_tester.prepare_config_and_inputs()
+        config.max_target_positions = 400
+        input_features = input_dict["input_features"]
+        model = TFWhisperForConditionalGeneration(config)
+        model.generate(input_features)
+        model.generate(input_features, num_beams=4, do_sample=True, early_stopping=False, num_return_sequences=3)
```
Comment (gante) [inline tests/models/whisper/test_modeling_tf_whisper.py:334]: I would skip this test for now, HF+TF+16bit support is very low atm

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 3]
```diff
@@ -0,0 +1,866 @@
...
+                [self.model_tester.num_attention_heads, subsampled_encoder_seq_length, subsampled_encoder_key_length],
+            )
+
+    def test_resize_token_embeddings(self):
+        # Overwritten method from parent; see `test_resize_embeddings_untied`
+        pass
+
+    def test_resize_tokens_embeddings(self):
+        # see `test_resize_embeddings_untied`
+        pass
```
Comment (gante) [inline tests/models/whisper/test_modeling_tf_whisper.py:499]: (duplicate?)

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 4]
```diff
@@ -0,0 +1,866 @@
...
+
+        EXPECTED_TRANSCRIPT = " Kimura san ni denwa wo kaite moraimashita"
+        self.assertEqual(transcript, EXPECTED_TRANSCRIPT)
+
+        model.config.forced_decoder_ids = processor.get_decoder_prompt_ids(language="ja", task="translate")
+        generated_ids = model.generate(input_features, do_sample=False)
+        transcript = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
+
+        EXPECTED_TRANSCRIPT = " I borrowed a phone from Kimura san"
+        self.assertEqual(transcript, EXPECTED_TRANSCRIPT)
```
Comment (gante) [inline tests/models/whisper/test_modeling_tf_whisper.py:866]: Missing in an ideal world:
- `test_large_batched_generation` (from PT test file)
- `test_tiny_en_batched_generation` (from PT test file)
- a copy of a small test, but that runs with XLA

**Review by gante — COMMENTED** [review]
🔥 🔥 🔥 

I have a bunch of `[XLA Notes]` comments -- they don't need to be addressed now. I've wrote them down as potential sources of problems for XLA :D

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 5]
```diff
@@ -0,0 +1,1353 @@
...
+        input_features=None,
+        head_mask=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+        training=False,
+    ):
+        r"""
+        Args:
+            input_features (`tf.Tensor` of shape `(batch_size, sequence_length, feature_size)`):
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/modeling_tf_whisper.py:636]: Same as above.

File: src/transformers/models/rag/modeling_tf_rag.py [hunk 6]
```diff
@@ -1262,6 +1262,8 @@ def extend_enc_output(tensor, num_beams=None):
                 eos_token_id=eos_token_id,
                 forced_bos_token_id=None,
                 forced_eos_token_id=None,
+                supress_tokens=None,  # Not used for RAG, were added after the whisper pull request
+                begin_supress_tokens=None,  # Not used for RAG, were added after the whisper pull request
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/rag/modeling_tf_rag.py]: They are now :) This is a hangover from an older version of the PT model. They've been ✂️

File: tests/models/whisper/test_modeling_tf_whisper.py (hunk 2 above)
Comment (ArthurZucker) [reply tests/models/whisper/test_modeling_tf_whisper.py:333]: But would have been good on PT side

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 7]
```diff
@@ -0,0 +1,1352 @@
...
+        # If the user passed a tuple for encoder_outputs, we wrap it in a BaseModelOutput when return_dict=True
+        elif return_dict and not isinstance(encoder_outputs, BaseModelOutput):
+            encoder_outputs = BaseModelOutput(
+                last_hidden_state=encoder_outputs[0],
+                hidden_states=encoder_outputs[1] if len(encoder_outputs) > 1 else None,
+                attentions=encoder_outputs[2] if len(encoder_outputs) > 2 else None,
+            )
+
+        # downsample encoder attention mask
+        # decoder_attention_mask = None
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py]: Yep! done 👍

**Review by ArthurZucker — COMMENTED** [review]
LGTM other than maybe XLA and the shared embedding

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 8]
```diff
@@ -0,0 +1,1352 @@
...
+        >>> inputs = processor(
+        ...     ds[0]["audio"]["array"], sampling_rate=ds[0]["audio"]["sampling_rate"], return_tensors="tf"
+        ... )
+        >>> input_features = inputs.input_features
+
+        >>> generated_ids = model.generate(inputs=input_features)
+
+        >>> transcription = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
+        >>> transcription
+        ' The quilter is the apostle of the middle classes and we are glad to welcome his'
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py:1271]: I think that's an accidental deletion from me. I'll update.

File: src/transformers/generation_tf_logits_process.py [hunk 9]
```diff
@@ -504,3 +504,64 @@ def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.
...
+        )
+        return scores
+
+
+class TFForceTokensLogitsProcessor(TFLogitsProcessor):
+    r"""This processor can be used to force a list of tokens. The processor will set their log probs to `0` and all
+    other tokens to `-inf` so that they are sampled at their corresponding index."""
+
+    def __init__(self, force_token_map):
+        self.force_token_map = dict(force_token_map)
```
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/generation_tf_logits_process.py:554]: I'm almost certain that anything like `get()` or a dictionary lookup will be done once when the function is traced, with any arguments treated as constants, rather than in each loop iteration with the arguments treated as variables.

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 10]
```diff
@@ -0,0 +1,1356 @@
...
+        self.layerdrop = config.encoder_layerdrop
+
+        self.embed_dim = config.d_model
+        self.num_mel_bins = config.num_mel_bins
+        self.padding_idx = config.pad_token_id
+        self.max_source_positions = config.max_source_positions
+        self.embed_scale = math.sqrt(self.embed_dim) if config.scale_embedding else 1.0
+
+        self.conv1 = tf.keras.layers.Conv1D(self.embed_dim, kernel_size=3, strides=1, padding="valid", name="conv1")
+        self.conv2 = tf.keras.layers.Conv1D(self.embed_dim, kernel_size=3, strides=2, padding="valid", name="conv2")
```
Comment (Rocketknight1) [inline src/transformers/models/whisper/modeling_tf_whisper.py:613]: Since this one took a second for me to figure out on review, it might deserve a comment!
```suggestion

        # Padding is added in call() to match the PyTorch implementation
        self.conv1 = tf.keras.layers.Conv1D(self.embed_dim, kernel_size=3, strides=1, padding="valid", name="conv1")
        self.conv2 = tf.keras.layers.Conv1D(self.embed_dim, kernel_size=3, strides=2, padding="valid", name="conv2")
```

File: tests/models/whisper/test_modeling_tf_whisper.py (hunk 4 above)
Comment (gante) [reply tests/models/whisper/test_modeling_tf_whisper.py:865]: 👍

**Review by Rocketknight1 — APPROVED** [review]
I focused on the core model code and comparing it to the PT implementation, since it seemed like @gante was handling generation and XLA compatibility. In the one place where I thought I'd found an error (missing padding in the `Conv1Ds`) that was being handled in the `call()`. I added a suggestion for a comment there, but other than that the core model code LGTM!

**Review by gante — COMMENTED** [review]
Correction to the forced ids logits processor test :)

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 11]
```diff
@@ -0,0 +1,1356 @@
...
+        input_features=None,
+        head_mask=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+        training=False,
+    ):
+        r"""
+        Args:
+            input_features (`tf.Tensor` of shape `(batch_size, sequence_length, feature_size)`):
```
Comment (amyeroberts) [inline src/transformers/models/whisper/modeling_tf_whisper.py:636]:
```suggestion
            input_features (`tf.Tensor` of shape `(batch_size, feature_size, sequence_length)`):
```

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 12]
```diff
@@ -0,0 +1,1352 @@
...
+        elif inputs_embeds is not None:
+            input_shape = tf.shape(inputs_embeds)[:-1]
+        else:
+            raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
+
+        # past_key_values_length
+        past_key_values_length = past_key_values[0][0].shape[2] if past_key_values is not None else 0
+
+        if inputs_embeds is None:
+            inputs_embeds = self.embed_tokens(input_ids)
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py:892]: Added 👍

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 13]
```diff
@@ -0,0 +1,1352 @@
...
+        head_mask=None,
+        decoder_head_mask=None,
+        cross_attn_head_mask=None,
+        encoder_outputs=None,
+        past_key_values=None,
+        decoder_inputs_embeds=None,
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py:1014]: Yep, added to the other layers now 👍

File: src/transformers/generation_tf_logits_process.py [hunk 14]
```diff
@@ -504,3 +504,64 @@ def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.
...
+        self.begin_suppress_tokens = list(begin_suppress_tokens)
+        self.begin_index = begin_index
+
+    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.Tensor:
+        scores = tf.cond(
+            tf.equal(cur_len, self.begin_index),
+            lambda: tf.tensor_scatter_nd_update(
+                scores,
+                indices=[[i, token] for i in range(scores.shape[0]) for token in self.begin_suppress_tokens],
+                updates=[-float("inf") for _ in range(scores.shape[0] * len(self.begin_suppress_tokens))],
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation_tf_logits_process.py:526]: FYI post XLA fixing: this wasn't a problem 👍

File: src/transformers/generation_tf_logits_process.py [hunk 15]
```diff
@@ -504,3 +504,64 @@ def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.
...
+    are not sampled."""
+
+    def __init__(self, suppress_tokens):
+        self.suppress_tokens = list(suppress_tokens)
+
+    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.Tensor:
+        scores = tf.tensor_scatter_nd_update(
+            scores,
+            indices=[[i, token] for i in range(scores.shape[0]) for token in self.suppress_tokens],
+            updates=[-float("inf") for _ in range(scores.shape[0] * len(self.suppress_tokens))],
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation_tf_logits_process.py:544]: FYI post XLA fixing: this wasn't a problem 👍

File: src/transformers/generation_tf_logits_process.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/generation_tf_logits_process.py:554]: FYI post XLA fixing: this was indeed a problem 😱

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 16]
```diff
@@ -0,0 +1,1352 @@
...
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    def _prepare_decoder_attention_mask(self, attention_mask, input_shape, past_key_values_length):
+        # create causal mask
+        # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+        batch_size, seq_len = input_shape[0], input_shape[1]
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/models/whisper/modeling_tf_whisper.py:766]: FYI post XLA fixing: this wasn't a problem 👍

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 17]
```diff
@@ -0,0 +1,1353 @@
...
+            cross_attn_head_mask=cross_attn_head_mask,
+            past_key_values=past_key_values,
+            decoder_inputs_embeds=decoder_inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+        # Decoder and encoder embeddings are tied
+        lm_logits = tf.matmul(outputs[0], self.model.get_input_embeddings().weights, transpose_b=True)
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py]: It looks like the `TFSharedEmbedding` layer is [flagged for being deleted](https://github.com/huggingface/transformers/blob/f0b490151e4a851c0821e1f215bb7a26565d24f7/src/transformers/modeling_tf_utils.py#L2817)  cc @gante 

I've tidied up the call a little bit though. Let me know what you think.

**Review by gante — APPROVED** [review]

File: src/transformers/models/whisper/modeling_whisper.py [hunk 18]
```diff
@@ -1102,16 +1102,16 @@ def get_encoder(self):
...
     def get_output_embeddings(self):
         return self.proj_out
 
     def set_output_embeddings(self, new_embeddings):
         self.proj_out = new_embeddings
 
+    def resize_token_embeddings(self, new_num_tokens: int) -> nn.Embedding:
+        new_embeddings = super().resize_token_embeddings(new_num_tokens)
+        return new_embeddings
+
```
Comment (sgugger) [inline src/transformers/models/whisper/modeling_whisper.py:1114]: Why the reordering? :thinking:

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding the port so quickly :muscle:

File: src/transformers/models/whisper/modeling_whisper.py (hunk 18 above)
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_whisper.py:1114]: I think this was a hangover when I was comparing the two and changed the PT model, rather than TF accidentally. I'll change it back 👍

**Review by patrickvonplaten — APPROVED** [review]
Very nice!

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 19]
```diff
@@ -0,0 +1,1403 @@
...
+    def get_encoder(self):
+        return self.model.encoder
+
+    def get_decoder(self):
+        return self.model.decoder
+
+    def decoder(self):
+        return self.model.decoder
+
+    def encoder(self):
```
Comment (patrickvonplaten) [inline src/transformers/models/whisper/modeling_tf_whisper.py:1128]: not really relevant for this PR but why is there both a `encoder` and a `get_encoder`  function?

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 20]
```diff
@@ -0,0 +1,1403 @@
...
+        )
+        self.embed_positions = TFWhisperPositionalEmbedding(
+            self.max_target_positions, config.d_model, name="embed_positions"
+        )
+
+        self.decoder_layers = [TFWhisperDecoderLayer(config, name=f"layers.{i}") for i in range(config.decoder_layers)]
+
+        self.layer_norm = tf.keras.layers.LayerNormalization(epsilon=1e-5, name="layer_norm")
+
+        self.gradient_checkpointing = False
```
Comment (patrickvonplaten) [inline src/transformers/models/whisper/modeling_tf_whisper.py:757]:
```suggestion
```
Think this is a copy-paset from PT

**Review by patrickvonplaten — APPROVED** [review]
Super nice!

File: src/transformers/models/whisper/modeling_tf_whisper.py (hunk 20 above)
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py]: Removed 👍

File: src/transformers/models/whisper/modeling_tf_whisper.py (hunk 19 above)
Comment (amyeroberts) [reply src/transformers/models/whisper/modeling_tf_whisper.py:1126]: Good question. Tbh, I was just copying this to match the PT model and didn't think about it. Looking at other models e.g. bart it seems to be a common pattern in the codebase.
